src/macos/mod.rs

```diff
@@ -47,56 +41,43 @@ impl PlatformManager {
 
         let cbc = callback.clone();
 
-        let thread = RunLoop::new_with_timeout(
-            move |alive| {
-                let mut devices = DeviceMap::new();
-                let monitor = try_or!(Monitor::new(), |e| callback.call(Err(e)));
-                let mut matches = KeyHandleMatcher::new(&key_handles);
-
-                'top: while alive() && monitor.alive() {
-                    for event in monitor.events() {
-                        devices.process_event(event);
-                    }
-
-                    // Query newly added devices.
-                    matches.update(devices.iter_mut(), |device, key_handle| {
-                        u2f_is_keyhandle_valid(device, &challenge, &application, key_handle)
-                            .unwrap_or(false /* no match on failure */)
-                    });
-
-                    // Iterate all devices that don't match any of the handles
-                    // in the exclusion list and try to register.
-                    for (path, device) in devices.iter_mut() {
-                        if matches.get(path).is_empty() {
-                            if let Ok(bytes) = u2f_register(device, &challenge, &application) {
-                                callback.call(Ok(bytes));
-                                return;
-                            }
-                        }
-
-                        // Check to see if monitor.events has any hotplug events that we'll need
-                        // to handle
-                        if monitor.events().size_hint().0 > 0 {
-                            debug!("Hotplug event; restarting loop");
-                            continue 'top;
-                        }
-                    }
-
-                    thread::sleep(Duration::from_millis(100));
+        // Start a new "sign" transaction.
+        let transaction = Transaction::new(timeout, cbc.clone(), move |device_ref, rx, alive| {
+            // Create a new device.
+            let dev = &mut Device::new(device_ref, rx);
+
+            // Try initializing it.
+            if !u2f_init_device(dev) {
+                return;
+            }
+
+            // Iterate the exlude list and see if there are any matches.
+            // Abort the state machine if we found a valid key handle.
+            if key_handles.iter().any(|key_handle| {
```

Comment:
Nit: feels like you should use the same chaining style here as around line 105... either's fine, but ... one. :)

Reply:
They're not quite the same. For `register()` we want to bail out as soon as we match one of the entries in the exclusion list. In `sign()` we want to actually get a list of all matching tokens and try to sign with those one by one.

I'll leave this as is and land it. If you disagree, let's open an issue and I'll be happy to address that in a follow-up :)